### DISCUSSION TITLE
How to access integer key/value in CBOR data?

### DISCUSSION THREAD

**Discussion by AachoLoya** [nlohmann/json#2627] [Q&A]
I'm loading CBOR data with schema 

```
{
   "field1":"value1",
   "field2": 
   {  1234:5678,
       2345:6789
    }
}
```

I'm not interested in JSON compatibility, hence keys in field2 are Integers - to keep the output size as small as possible.

Found the following code to parse over field2 items: 

```cpp
for (auto& el : data["field2"].items())
{
  cout << "Key:" << el.key() << " Value: " << el.value();
}
```

but key() and value() both return string. Does nlohmann/json natively store all data as string? If not, then how do I access the actual integer key & value?

**Comment by nlohmann** [comment]
This seems to be related to #2477.

**Reply by AachoLoya** [reply]
looks like pull request is just sitting there due to some issues? Any idea when it'll be part of a release?

**Reply by nlohmann** [reply]
I am currently hesitant to merge, because I did not see the need, and this seemed to be a MessagePack-only issue. But now I see that also CBOR uses integer keys.

However, even with an extended SAX interface, the object keys stored in a `basic_json` value would remain to be strings.

**Reply by AachoLoya** [reply]
This would have been a really nice feature, since I'm already using this header/library in my code. Also would be needed to fully support conversion from CBOR. I guess for now, I'll look for some other library.

**Reply by nlohmann** [reply]
Yes, it would be nice. But how would this library store integer keys?

**Reply by gregmarr** [reply]
Maybe add `binary_map` to `value_t` that is a cross between `array` and `object`.  It means that the data is in `json_value::array` but in pairs of key and value.  I know it's not trivial, but that seems like the only way that doesn't require adding more data members.

**Reply by AachoLoya** [reply]
#2477 Doesn't his change already address the issue of integer based keys?

**Reply by gregmarr** [reply]
@AachoLoya Looks like #2477 has a few restrictions.
1. The key is converted to a string when it is read.
2. The information about whether the key was an integer or a string in the file is lost.
3. Other valid types besides integers and strings are not supported as keys.
4. There is no support for writing out files with non-string keys.

If your use case works with those restrictions, then that's probably enough.

**Reply by AachoLoya** [reply]
Just to confirm, currently this library won't even parse integer keys? or would it parse and store them as strings?

**Reply by gregmarr** [reply]
My reading of the PR is that it doesn't parse them now, but #2477 adds the capability to parse and store as string.

**Reply by AachoLoya** [reply]
@nlohmann Can we at least get support for int etc keys as strings? This header/library is very useful and I'd prefer to stick with this.
